File: Data Structures & Algorithms/search-2d-matrix/submission-16.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
          int left = 0;
          int right = matrix.length -1;

          while(left <= right){
            int middle = (left + right) / 2;
            System.out.println(middle);
            int [] current = matrix[middle];

            if(target < current[0]){
                right--;
            }else if(target > current[current.length -1]){
                left++;
            }else {
                return isTargetInArray(current, target);
            }
          }
        
        return false;
    }

    public boolean isTargetInArray(int [] nums, int target){
        int left = 0;
        int right = nums.length -1;

        while(left <= right){
            int middle = (left + right) / 2;

            if(nums[middle] > target){
                right--;
            }else if(nums[middle] < target){
                left++;
            }else {
                return true;
            }
        }

        return false;
    }
}

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-16.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        result = float('inf')

        left = 0
        right = len(nums) - 1

        while left <= right:

            if nums[left] < nums[right]:
                result = min(result, nums[left])

            middle = (left + right) // 2

            result = min(nums[middle], result)

            print(nums[middle])

            if nums[middle] >= nums[left]:
                print("we move left")
                left = middle + 1
            else:
                print("we move right")
                right = middle - 1
        
        return result
